### PR TITLE
#Hash in URL scrolls directly to card

### LINKED ISSUE #1753: Scroll position slightly off for really long madlibs
**Describe the bug**
Examples of problems when sticking madlib header contains really long conditions / place names and needs more vertical space; the scrolling is a little off. This is a pretty rare edge case that doesn't break anything just doesn't look as spiffy so deprioritizing this fix. 

**To Reproduce**
1. Go to `/exploredata?mls=1.substance-3.02198&mlp=disparity&dt1=covid_hospitalizations&demo=race_and_ethnicity&dt2=covid_hospitalizations`

**Expected behavior**
Scroll position and menu placement should be dynamically calculated off the header height, to keep it looking pretty. 

**Screenshots**
Screen Shot 2022-08-16 at 9 46 10 AM
Screen Shot 2022-08-16 at 9 45 47 AM

### PR DESCRIPTION
**Pull request by benhammondmusic**
## Preview Link

<a href="https://deploy-preview-1761--health-equity-tracker.netlify.app/exploredata#data-table" target="_blank">
  <br />
Click to demo updates </a> 

## Description

- updates display text on table of contents 
- appending a particular hash to the url (like https://healthequitytracker.org/exploredata#data-table) will automatically scroll  to the relevant card
- to account for long load times and hard to predict card heights, this will set up an automated scroll that repeats every 100ms until a user interacts by moving the scroll wheel, pressing any key, or clicking/touching the screen. As a fallback, the auto-scroll will cancel after a set time period
-  because the header can contain 1,2 or 3 lines of text, and has different margins and features based on the "sticky" setting, this will measure the header element directly and adjust the scroll-margin amounts so that the card scrolls into actual view and is not obscured by the header; same with the vertical placement of the floating table of contents menu
- remove scroll-margin for mobile since no sticky header
- the presence of a #hash link will automatically suppress the warm welcome
- previously any changes to the madlib would trigger a "scroll to top". however to prevent multiple competing scrollings, now it scrolls to top specifically on changes to the location or condition dropdown, and on changes to the madlib carousel mode
- clicks on map locations or map card breadcrumbs will keep clicked card aligned in view by appending the appropriate #hash (for unknown or rate map)
- adds link icons to each card that copy that specific link to the clipboard, and alerts the user the link was copied
- refactors lots of older ReactRouter specific links into normal `<a>` links. the fancy ones were having trouble with forward/back functionality; the normal html seems to be working great
-  clicking table of contents needs to change focus to clicked card. Used tabIndex -1 which allows us to shift focus programmatically but doesn't place card in the normal tab order which is the expected a11y behavior
- removes older way of handling "scroll to definitions" and "scroll to missing data" and uses same `exploredata#definitions-missing-data` hash linking instead. removes a lot of unneeded prop drilling

## Motivation and Context

fixes #1753

## How has this been tested?

- extensive manual testing on various devices
- added a playwright test the ensures clicking the table of contents link: scrolls the correct card into view, updates the url hash, and focuses the correct card

- [x] macOS chrome
- [x] macOS safari
- [x] macOS firefox
- [x] macOS edge
- [x] ios safari
- [ ] android? 
- [ ] windows?

## Screenshots (if appropriate):

Link Copy to Clipboard working

https://user-images.githubusercontent.com/41567007/190525355-6fd68007-d787-4fc2-abfb-4407422b5c66.mov

Demo on various cards / tracker modes

https://user-images.githubusercontent.com/41567007/190509943-20e1b57e-2900-4d42-9e95-159892814e8a.mp4

Link directly to age-adjust card from age-adjust methodology tab

https://user-images.githubusercontent.com/41567007/190508978-fbd9cbc1-b4d2-454d-ad71-eca79c4cb59d.mp4

Clicking on Maps and Breadcrumbs keeps correct card in focus

https://user-images.githubusercontent.com/41567007/190511891-7ed9dc2b-5ef3-49b4-999d-e81ae4253f16.mov

Snackbars over various backgrounds

Screen Shot 2022-09-20 at 2 18 24 PM
Screen Shot 2022-09-20 at 2 18 34 PM

## Types of changes

- New feature (non-breaking change which adds functionality or content)

### CODE REVIEW

File: frontend/src/pages/ExploreData/ExploreDataPage.tsx [hunk 1]
```diff
@@ -128,6 +124,10 @@ function ExploreDataPage() {
   if (params[SHOW_ONBOARDING_PARAM] === "false") {
     showOnboarding = false;
   }
+
+  // if there is an incoming #hash; bypass the warm welcome entirely
+  if (location.hash !== "") showOnboarding = false;
```
Comment (eriwarr) [inline frontend/src/pages/ExploreData/ExploreDataPage.tsx:129]: Nice!

**Review by eriwarr — APPROVED** [review]
Looks awesome! Great job Ben!

File: frontend/src/cards/ui/CopyLinkButton.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import React, { useState } from "react";
+import LinkIcon from "@material-ui/icons/Link";
+import styles from "./CopyLinkButton.module.scss";
+import { IconButton, Snackbar } from "@material-ui/core";
+import { ScrollableHashId } from "../../utils/hooks/useStepObserver";
+// import { useLocation } from "react-router-dom";
```
Comment (joshzarrabi) [inline frontend/src/cards/ui/CopyLinkButton.tsx:6]: delete this line?

File: frontend/src/cards/ui/CopyLinkButton.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        aria-label={`copy direct link to: ${cardName}`}
+        onClick={handleClick}
+      >
+        <LinkIcon />
+      </IconButton>
+      <Snackbar
+        open={open}
+        autoHideDuration={5000}
+        onClose={handleClose}
+        // onClick={handleClose}
```
Comment (joshzarrabi) [inline frontend/src/cards/ui/CopyLinkButton.tsx:44]: delete this line?

**Review by joshzarrabi — APPROVED** [review]
This is huge, thank you so much.

**Pull request merged**